### DISCUSSION TITLE
Feature request: $cache

### DISCUSSION THREAD

**Discussion by j-greig** [alpine-collective/alpine-magic-helpers#90] [Ideas]
For use with $fetch :)

**Comment by SimoTod** [comment]
Thanks for suggesting it @j-greig.
Could you post an example of how you immagine the feature to work?

**Reply by SimoTod** [reply]
I would immagine it as
`$cache(expression)` and `$cache(key, expression)`. Both will return the value of `expression` the first time (it could be a function, the result of another magic helper like `$fetch` or even a simple variable), cache it and return the same value the following times.
The version without key will just cache the value for the current directive (despite the logic being different, it achieves the same result as v-once in vue).
The version with the key will allow to share the cache between different directives so the first one will compute the value and the second one will just use the cached result.

```html
<div x-data="...">
    <p x-text="$cache(foo())> <!-- foo will be called just one time -->
</div>
```
And
```html
<div x-data="...">
    <p x-text="$cache('bar', foo())> <!-- foo will be called just one time -->
    <p x-text="$cache('bar', foo())> <!-- This element will always use the cached value  -->
</div>
<div x-data="...">
    <p x-text="$cache('bar', foo())> <!-- TBC if we want to support intra components cache, probably we do-->
</div>
```

**Reply by KevinBatdorf** [reply]
What about a third parameter `seconds` and we clear it out after n seconds?

**Reply by j-greig** [reply]
Hello, I'm not good enough at JS to really suggest how this would work 😂

But you can see the example that @KevinBatdorf was helping me with today below, I guess this is the kind of scenario where I would imagine using this... https://codepen.io/KevinBatdorf/pen/XWNpvrm

Cheers 👍

**Reply by SimoTod** [reply]
It's going to be cached at page level so we are talking about short time frames (it won't persist between different pages), probably seconds is not needed but let's see. 👍

**Reply by KevinBatdorf** [reply]
I could see myself using it with an expires option in the example @j-greig was working on actually. Where you have a dataset that you probably only need to fetch when the user interacts, but only every n minutes. Useful on pages you leave open long term (I have a few tabs open 24/7) or maybe Electron apps (not sure how Alpine runs in Election though)

Alternatively, and probably better, a separate `$onUpdate()` callback that fires on any update with a callback. 
Think `$watch($el, () => {})` for example.

Again, just brainstorming right now

**Reply by KevinBatdorf** [reply]
Another thing we could probably solve from that demo is the `x-for` + `$parent` relationship.

I used `x-init="$nextTick(() => loadCoin($parent.coins[$el.__x_for_key]))"` to grab the outer component data after the key is set, which isn't very user friendly.

**Reply by SimoTod** [reply]
Yeah, the demo seems a bit complex because of the Alpine limitation when using x-for and nested x-data. There are simpler ways without using nested x-data but it's not the point of the thread.
I was just worried that a lot of users will get confused if we provide too many options and things to configure (for example, what would it be a sensible value for seconds? 10? 30? 60? It's probably better to have something to invalidate the cache so users can use it on actions or watcher like your last example if they really need to)

**Reply by KevinBatdorf** [reply]
For `$cache` yeah let's see what you come up with first. I've been working with Laravel Cache this weekend and trying various warming strategies so that api is fresh in my head. ie.`Cache::put('key', 'value', $seconds = 10);`

But I would probably have it default to "forever" for our helper so there's no overhead, and just give the user the option to override it. We can add that later though if there's demand for it

**Reply by SimoTod** [reply]
@j-greig in the meantime, this example should do the same trick without calling fetch every time: https://codepen.io/SimoTod/pen/YzpZPMJ

**Reply by KevinBatdorf** [reply]
That's a good optimization. Makes the naming for a helper for `$el.__x_for_key` more obvious as `$currentKey` too, which would replace the title in that demo: `<p x-text="$el.__x_for_key"></p>`

updated: https://codepen.io/KevinBatdorf/pen/XWNpvrm

We'd have to wait for it to be available though too.

**Comment by SimoTod** [comment]
Moved to discussions since it's not an actual bug but a feature request.
